Relation bug?

```
def test_logic():
    from logpy import run,var,Relation,fact
    valido = Relation()
    fact(valido,"a","b")
    fact(valido,"b","a")
    x = var()
    assert run(0,x, valido(x,x)) == ()

E       assert ('a', 'b') == ()
E         Left contains more items, first extra item: 'a'
```

--- a/logpy/facts.py
+++ b/logpy/facts.py
@@ -29,8 +29,28 @@ class Relation(object):
             self.index[key].add(fact)
 
     def __call__(self, *args):
-        def f(s):
-            args2 = reify(args, s)
+        """ Returns an evaluated (callable) goal, which returns a list of
+        substitutions which match args against a fact in the knowledge base.
+
+        *args: the goal to evaluate. This consists of vars and values to
+               match facts against.
+
+        >>> from logpy.facts import Relation
+        >>> from logpy.variable import var
+        >>>
+        >>> x, y = var('x'), var('y')
+        >>> r = Relation()
+        >>> r.add_fact(1, 2, 3)
+        >>> r.add_fact(4, 5, 6)
+        >>> list(r(x, y, 3)({})) == [{y: 2, x: 1}]
+        True
+        >>> list(r(x, 5, y)({})) == [{y: 6, x: 4}]
+        True
+        >>> list(r(x, 42, y)({}))
+        []
+        """
+        def goal(substitution):
+            args2 = reify(args, substitution)
             subsets = [self.index[key] for key in enumerate(args)
                                        if  key in self.index]
             if subsets:     # we are able to reduce the pool early
@@ -44,9 +64,9 @@ class Relation(object):
             assert not any(var in s for var in vars)
 
             return (merge(dict(zip(vars, index(fact, varinds))), s)
-                              for fact in self.facts
+                              for fact in facts
                               if vals == index(fact, valinds))
-        return f
+        return goal
 
     def __str__(self):
         return "Rel: " + self.name

--- a/logpy/facts.py
+++ b/logpy/facts.py
@@ -1,6 +1,8 @@
-from .util import intersection, index
-from .core import conde, reify, isvar
+from .core import reify
+from .unification import unify
+from .util import intersection
 from toolz import merge
+
 
 class Relation(object):
     _id = 0
@@ -57,15 +59,12 @@ class Relation(object):
                 facts = intersection(*sorted(subsets, key=len))
             else:
                 facts = self.facts
-            varinds = [i for i, arg in enumerate(args2) if isvar(arg)]
-            valinds = [i for i, arg in enumerate(args2) if not isvar(arg)]
-            vars = index(args2, varinds)
-            vals = index(args2, valinds)
-            assert not any(var in s for var in vars)
 
-            return (merge(dict(zip(vars, index(fact, varinds))), s)
-                              for fact in facts
-                              if vals == index(fact, valinds))
+            for fact in facts:
+                unified = unify(fact, args2, substitution)
+                if unified != False:
+                    yield merge(unified, substitution)
+
         return goal
 
     def __str__(self):

--- a/logpy/tests/test_facts.py
+++ b/logpy/tests/test_facts.py
@@ -33,3 +33,31 @@ def test_fact():
     assert (2, 3) in rel.facts
     assert (3, 4) in rel.facts
 
+def test_unify_variable_with_itself_should_not_unify():
+    # Regression test for https://github.com/logpy/logpy/issues/33
+    valido = Relation()
+    fact(valido, "a", "b")
+    fact(valido, "b", "a")
+    x = var()
+    assert run(0,x, valido(x,x)) == ()
+
+def test_unify_variable_with_itself_should_unify():
+    valido = Relation()
+    fact(valido, 0, 1)
+    fact(valido, 1, 0)
+    fact(valido, 1, 1)
+    x = var()
+    assert run(0,x, valido(x,x)) == (1,)
+
+def test_unify_tuple():
+    # Tests that adding facts can be unified with unpacked versions of those
+    # facts.
+    valido = Relation()
+    fact(valido, (0, 1))
+    fact(valido, (1, 0))
+    fact(valido, (1, 1))
+    x = var()
+    y = var()
+    assert set(run(0, x, valido((x, y)))) == {0, 1}
+    assert set(run(0, (x, y), valido((x, y)))) == {(0, 1), (1, 0), (1, 1)}
+    assert run(0,x, valido((x,x))) == (1,)

--- a/logpy/tests/test_facts.py
+++ b/logpy/tests/test_facts.py
@@ -58,6 +58,6 @@ def test_unify_tuple():
     fact(valido, (1, 1))
     x = var()
     y = var()
-    assert set(run(0, x, valido((x, y)))) == {0, 1}
-    assert set(run(0, (x, y), valido((x, y)))) == {(0, 1), (1, 0), (1, 1)}
+    assert set(run(0, x, valido((x, y)))) == set([0, 1])
+    assert set(run(0, (x, y), valido((x, y)))) == set([(0, 1), (1, 0), (1, 1)])
     assert run(0,x, valido((x,x))) == (1,)

